README
plus4emu 1.2.11 beta
====================

plus4emu is an open source, portable emulator of the Commodore Plus/4
computer, written in C++, and supporting Windows and POSIX platforms
(32 bit Windows, 32 and 64 bit Linux, and MacOS X have been tested).
It implements accurate, high quality hardware emulation, however, the
system requirements are higher than that of most other emulators.

Features
========

Plus/4 emulation
----------------

  * cycle exact emulation of the 7501/8501 CPU, with full support for
    all documented and undocumented opcodes
  * cycle exact TED 7360/8360 emulation with very high level of
    compatibility, implements all documented video modes, PAL/NTSC mode
    switching, accurate sound emulation, as well as all known display
    artifacts; all registers can be read and written, and the internal
    operation and timing is reproduced with very good accuracy
  * RAM size can be 16, 32, 64, 256, or 1024 kilobytes (the latter two
    using the Hannes memory expansion)
  * up to 8 16 kilobyte banks of ROM can be loaded from external image
    files; using custom ROM files for the emulated floppy drives and
    printer is also possible
  * limited emulation of the ACIA 6551 registers
  * tape emulation with playback, recording, and setting tape position;
    markers can be created for quick positioning to specific tape
    locations (useful for tapes with multiple files); uses custom file
    format which is PCM audio data with 1 to 8 bits of sample resolution
    and variable sample rate, and header including the table of markers;
    there is also limited (read only) support for the commonly used MTAP
    file format, as well as read-write (although without markers)
    support for sound files like WAV, AIFF, etc.
  * hardware level emulation of the 1541, 1551, and 1581 floppy disk
    drives, with support for D64 and D81 files; compatibility with 1541
    and 1551 fast loaders is fairly good in recent versions of the
    emulator
  * IEC level 1551 floppy drive emulation for accessing .PRG, .P00,
    .R00, .S00, and .U00 files on the PC filesystem; this also supports
    the C, I, N, P, R, S, and V DOS commands
  * light pen emulation (compatible with the Trojan Light Pen software)
  * IEC level MPS-801 printer emulation
  * hardware level 1526/MPS-802 printer emulation
  * load and automatically start .PRG, .P00, .D64, .D81, or .TAP file on
    start-up
  * .PRG, .P00, .D64, and .D81 files can be associated with plus4emu on
    Windows
  * SID chip (MOS 6581 and 8580) emulation using version 1.0 of the
    reSID library written by Dag Lem; the Digi-Blaster expansion by
    Solder is also supported

General
-------

  * graphical user interface using the FLTK library
  * software (FLTK based) or OpenGL video, with resizable emulator
    window, fullscreen mode, brightness/contrast/gamma can be set
    globally or separately for red/green/blue channels, color
    saturation control; additional features in OpenGL mode only: single
    or double buffered (with synchronization to vertical refresh) mode,
    linear texture filtering, and some display effects
  * high quality television (PAL and NTSC) emulation; implements full
    horizontal and vertical synchronization, blanking, equalization,
    sync, and burst signals, interlace, and (if OpenGL 2.0 shaders are
    available) separate luma/chroma filtering to emulate the blurredness
    of the TV picture
  * real time audio output uses the PortAudio library (v18 or v19), with
    support for many native audio APIs (MME/DirectSound/ASIO on Windows,
    OSS/ALSA/JACK on Linux, and CoreAudio on MacOS X); high quality
    sample rate conversion with low aliasing; volume control, two first
    order highpass filters with configurable cutoff frequency, and an
    optional parametric equalizer can be applied to the audio signal
  * recording audio output to a WAV format sound file
  * recording video and sound output to an AVI format video file, with
    RLE8 or uncompressed YV12 video at 24 to 60 frames per second, and
    48000 Hz 16-bit PCM audio
  * saving screenshots as 768x576 PNG format files
  * saving and loading snapshots of the state of the emulated machine
  * demo recording (snapshot combined with stream of keyboard events
    which can be played back with accurate timing)
  * GUI debugger with support for breakpoints/watchpoints, viewing the
    current state of CPU registers and memory paging, displaying memory
    dump or loading from or saving to a file in ASCII or binary format,
    and disassembler with support for all undocumented 6502 opcodes.
    The debugger also includes a TEDMON-like monitor with support for
    all the commands, as well as some improvements and additional
    features (assemble/disassemble offset, saving disassembly to a file,
    physical (22 bit) memory addressing mode, undocumented CPU opcodes
    in the assembler and disassembler, trace, search/replace pattern in
    memory, and more).
    It is also possible to debug the 6502 compatible CPUs in floppy
    drives and printers.
    The debugger supports scripting in the Lua language, to allow for
    advanced uses like breakpoints with custom defined, complex set of
    conditions.
  * configurable keyboard map for the emulated machine; it is also
    possible to use external game controller devices like joysticks and
    gamepads
  * mouse and clipboard support for setting the cursor position,
    copying, and pasting text in the emulated machine:
      left click: sets the cursor position
      right click: sets the cursor position, and copies the selected
        word to the clipboard
      ctrl + right click: sets the cursor position, and copies text from
        the selected line to the clipboard
      shift + right click: copies text from the whole screen to the
        clipboard without changing the cursor position
      middle click: sets the cursor position, and pastes text from the
        clipboard
      shift + middle click: pastes text from the clipboard without
        changing the cursor position
  * the Plus/4 emulation can be used in other programs as a shared
    library with a C interface; see the plus4lib/plus4emu.h header file
    for details, and also plus4lib/sample.c for a simple example of
    using the interface

The emulator package also includes some utilities:

  * (p4)compress - compresses Plus/4 .PRG and .P00 files with very good
    compression ratio (CA65 assembler source code for the decompressor
    is available)
  * (p4)makecfg - reinstalls the configuration files of the emulator
  * p4fliconv - converts JPEG and other format images to various Plus/4
    video modes (hires/multicolor, FLI, interlace, etc.); it can be used
    both from the command line, and as a GUI program which uses the
    emulator for accurate previews. CA65 assembler sources for the FLI
    viewer are also available.
  * p4sconv - converts PixelShop .P4S images to any of the output
    formats supported by p4fliconv
  * (p4)tapconv - converts MTAP format tape files to the native plus4emu
    tape format

Installation
============

Linux
-----

On Linux and other POSIX platforms, the emulator is installed from the
source code, available at the GitHub download page
  https://github.com/istvan-v/plus4emu/releases/
or the most recent state of the code can be downloaded from Git with the
following command:
  git clone https://github.com/istvan-v/plus4emu.git
In addition to the standard development tools (a recent version of the
GNU C/C++ compiler, binutils, etc.), you need the following packages:

  * SCons (http://www.scons.org/)
  * Python interpreter for running SCons
  * FLTK 1.3.x (http://www.fltk.org/software.php?VERSION=1.3.4) or
    1.1.x (http://www.fltk.org/software.php?VERSION=1.1.10)
    NOTES:
      * this library should be compiled with the --enable-threads
        'configure' option - many Linux distributions include binaries
        of the FLTK library built without --enable-threads, so you may
        need to compile it from sources
      * on MacOS X, FLTK 1.1.7 needs to be patched with the included
        fltk-1.1.7-MacOSX.patch file
  * PortAudio (http://www.portaudio.com/download.html), version 18 and
    19 are supported, but v19 is recommended
  * libsndfile (http://www.mega-nerd.com/libsndfile/#Download)
  * SDL (http://www.libsdl.org/) 1.2 for joystick input (optional);
    NOTE: on Linux, versions 1.2.10 and newer do not work if they are
    statically linked and were built with video support (--enable-video)
  * Lua (http://www.lua.org/ or http://www.luajit.org/) 5.1, 5.2, or 5.3
    for scripting in the debugger (optional)

Once these are installed, you can edit the file 'SConstruct' in the top
level source directory for setting compiler flags etc., and run the
command 'scons' for building the emulator. The resulting executable
files (plus4emu and the utilities listed above) can be copied to any
directory that is in the PATH; on MacOS X, an .app package is created in
'plus4emu.app'.
When installing the first time, you also need to set up configuration
files and ROM images:

  * copy all .rom files from the roms directory to ~/.plus4emu/roms (or,
    in the case of MacOS X, ~/Library/Application Support/plus4emu/roms)
  * after installing the ROM images, run 'plus4emu', and click OK to the
    windows that pop up asking for the base directory of configuration
    and data files, and if configuration files should be installed

It is possible to reinstall configuration files later by running the
'p4makecfg' utility.

Alternatively, the emulator and all required files can be installed
under ~/bin and ~/.local/share with the command 'scons install', which
also runs p4makecfg.

Windows
-------

A binary package with an installer is available at the GitHub download
page:
  https://github.com/istvan-v/plus4emu/releases/
Older (pre-1.2.10) versions can be downloaded from SourceForge:
  https://sourceforge.net/projects/plus4emu/files/
To install, just run the executable, and follow the instructions.
When asked if configuration files should be reinstalled, click 'OK'
when installing the first time, but this step can be skipped in later
installations to preserve the configuration.

Usage
=====

Command line options
--------------------

  -h
  -help
  --help
    print the list of available command line options
  -cfg <FILENAME>
    load an ASCII format configuration file on startup, and apply
    settings
  -prg <FILENAME>
    load and run program file on startup
  -disk <FILENAME>
    load and automatically start disk image on startup
  -tape <FILENAME>
    load and automatically start tape image on startup
  -snapshot <FILENAME>
    load snapshot or demo file on startup
  -keybuf <TEXT>
    type TEXT on startup (can be any length, or @FILENAME to read from a
    file)
  -opengl
    use OpenGL video driver (this is the default, and is recommended
    when hardware accelerated OpenGL is available)
  -no-opengl
    use software video driver; this is slower than OpenGL when used at
    high resolutions, and also disables many display effects, but should
    work on all machines; however, it will use a color depth of 24 bits,
    while in OpenGL mode the textures are 16 bit (R5G6B5) only, to
    improve performance
  -colorscheme <N>
    select GUI color scheme N (0, 1, 2, or 3)
  OPTION=VALUE
    set configuration variable 'OPTION' to 'VALUE'; the available
    variable names are the same as those used in ASCII format
    configuration files
  OPTION
    set boolean configuration variable 'OPTION' to true
  FILENAME
    load and automatically start a file with .prg, .p00, .d64, .d81, or
    .tap extension

'File' menu
-----------

Configuration / Load from ASCII file

  Select and load an ASCII format configuration file and apply the new
  settings. If the configuration file does not include all the supported
  options, those that are omitted are left unchanged.

Configuration / Load from binary file

  Load a plus4emu format binary file, which may be a previously saved
  snapshot or demo, a binary format configuration file, or a program
  file with plus4emu header.

Configuration / Save as ASCII file

  Save the current emulator configuration to an ASCII text file, which
  can be edited with any text editor, and can be loaded at a later time.

Configuration / Save

  Save the current emulator configuration in binary format to the
  default file (~/.plus4emu/plus4cfg.dat). This is also automatically
  done when exiting the emulator.

Configuration / Revert

  Reload emulator configuration from ~/.plus4emu/plus4cfg.dat, and apply
  the original settings.

Save snapshot

  Save a snapshot of the current state of the emulated machine to the
  selected file. The snapshot will also include the current memory
  configuration and ROM images, but clock frequency and timing settings
  are not restored when loading a snapshot. The file format may be
  subject to changes between different releases of the emulator. Note
  that the state of any floppy drives is currently not saved, and the
  drives are reset on loading a snapshot.

Load snapshot (F7)

  Load a plus4emu format binary file, which may be a previously saved
  snapshot or demo, a binary format configuration file, or a program
  file with plus4emu header.

Quick snapshot / Set file name

  Select file name for quick snapshots. The default is
  ~/.plus4emu/qs_plus4.dat. This setting is not saved on exit.

Quick snapshot / Save (Ctrl + F9)

  Save snapshot to the quick snapshot file (see notes above).

Quick snapshot / Load (Ctrl + F10)

  Load the quick snapshot file if it exists.

Record demo

  Save snapshot (including clock frequency and timing settings) and
  record keyboard events to the selected file until the recording is
  stopped. The events can be replayed with accurate timing when the
  file is loaded later. Note that the file format may change between
  different releases of the emulator, and the timing may also be
  incorrect when using a different version to play a demo file.

Stop demo (Ctrl + F12)

  Stop any currently running demo playback or recording.

Load demo (F7)

  Load a plus4emu format binary file, which may be a previously saved
  snapshot or demo, a binary format configuration file, or a program
  file with plus4emu header.

Record audio / Start...

  Write 16 bit signed PCM sound output to a WAV format sound file.

Record audio / Stop

  Close sound output file if it is currently being written.

Record video / Start...

  Open new AVI file for video recording. This increases the CPU usage
  significantly, and since the data is written without compression, it
  will take up a lot of disk space. If the size of the output file
  reaches 2 GB, it is automatically closed, and the emulator asks for
  a new file to continue the recording.
  NOTE: the video and audio streams in the AVI file are not affected by
  any of the display or sound configuration settings.

Record video / Stop

  Stop video capture, and close any AVI file that is currently being
  written.

Save screenshot (F6)

  Save a screenshot in PNG format. The video output is captured
  immediately after activating this menu item, and is saved at a
  resolution of 768x576 without any processing (color correction,
  effects, etc.).

Quit (Shift + F12)

  Exit the emulator.

'Machine' menu
--------------

Reset / Reset (F11)

  This has the same effect as using the reset button on the real
  machine.

Reset / Force reset (Ctrl + F11)

  In addition to a normal reset, make sure that the emulated machine is
  really restarted using the standard ROM reset routines, and do not
  allow programs to disable reset by setting custom (RAM) handlers;
  however, most of the RAM data is still preserved.
  This option will also turn off automatically enabled extensions that
  increase CPU usage, such as SID emulation and 1541/1551/1581 drives
  with no disk opened.

Reset / Reset clock frequencies

  Reset clock frequency and timing settings to those specified in the
  machine configuration; this is normally only useful after demo
  playback, which may override the settings. The SID and ACIA
  configuration, which may be changed by snapshot or demo loading, are

NEWS
Changes in version 1.2.11
-------------------------

  * PAL/NTSC emulation can now use custom GLSL shaders
  * added new disable unit 8/9 buttons to the disk configuration window
  * an empty D64 or D81 image is automatically created when trying to
    open a file that does not exist
  * disk and tape images can be opened from .zip format archives, the
    first packed file with matching extension is temporarily extracted
    and used in read-only mode
  * added support for comments (beginning with ; or #) in the breakpoint
    list
  * fixed DigiBlaster bug, SID external input was muted by default
  * fixed horizontal scroll bug in multicolor character mode
  * minor fixes in 1551 and TED emulation
  * fixed video breakpoints not working when the priority threshold is
    not zero

Changes in version 1.2.10.1
---------------------------

  * fixed preview bug in p4fliconv

Changes in version 1.2.10
-------------------------

  * new configuration options for C64 compatible SID emulation
  * updated SID emulation to use reSID 1.0 instead of 0.16
  * video display improvements and bug fixes; the screen is no longer
    blanked while the emulator is paused
  * the debugger window is hidden only after a short delay when it is
    closed, this fixes the flickering while using the step buttons
  * the editor buffer in the monitor has been increased from 120 to 160
    lines
  * memory dump is now also shown in ASCII format
  * new execute ('x') type of breakpoints, the ignore data reads button
    has been removed
  * in addition to the default hexadecimal format, it is also possible
    to specify numbers in the monitor in binary (e.g. %1010), octal
    (12o), and decimal (10l) format
  * new "step" buttons for stepping to the 16-bit address at the top of
    the stack ('Return') or any 16-bit address specified by the user
    ('Step to')
  * the debugger prints more detailed information about syntax and
    runtime errors in Lua scripts, and there is a new button to insert
    an empty breakpoint callback function
  * it is possible to delete single breakpoints in Lua scripts by using
    a negative priority value
  * window icons have been implemented on non-Windows platforms if the
    emulator is built with FLTK 1.3.3 or newer
  * the debugger window has been made resizeable within a limited range
  * audio output fixes and improvements, WASAPI is now supported on
    Windows
  * screenshots and printer output are saved in PNG format instead of
    TGA
  * various improvements and new features in the compress utility
  * snapshot and demo files can be compressed in epcompress format
  * added support for building with FLTK 1.3.x and Lua 5.2/5.3, and for
    building 64-bit Windows binaries
  * OpenGL has been made an optional dependency at compile time
  * minor bug fixes

Changes in version 1.2.9.2
--------------------------

  * minor bug fixes

Changes in version 1.2.9.1
--------------------------

  * some improvements and optimizations in the p4fliconv and compress
    utilities
  * a few minor bugs have been fixed

Changes in version 1.2.9
------------------------

  * added a C interface that allows using the emulator as a shared
    library in other programs; the library and header file can be found
    in the 'plus4lib' subdirectory of the binary packages, and a simple
    example program with compile instructions is also included
  * implemented IEC level MPS-801 printer emulation, in addition to the
    previously existing hardware level MPS-802 emulation
  * added a new button to the debugger GUI that makes it possible to
    break only when a branch is taken (i.e. the opposite of 'Step over')
  * more accurate PAL emulation in OpenGL mode when the display quality
    is set to 3
  * LED display is now also implemented for IEC level drive emulation
  * the ROM images required by the emulator are included with the binary
    packages, and no longer need to be downloaded separately
  * fixed a bug in the 'Step over' function of the debugger
  * fixed/improved support for D64 error information; this allows some
    copy protected games that previously did not work to load correctly
  * the 'compress' utility has two new compression methods, some new
    options, and the original method is 3-5 times faster now while
    producing identical files; also, compression support in 'p4fliconv'
    is improved somewhat
  * various minor bug fixes, optimizations, and other improvements

Changes in version 1.2.8.1
--------------------------

  * fixed file overwrite (@) bug in IEC level drive emulation

Changes in version 1.2.8
------------------------

  * new option for saving the printer output to a text file, either as
    the raw characters sent to the printer, or converted to ASCII
  * slightly improved compression when saving the printer output in TGA
    format
  * added Windows icons from the Plus/4 icon set by SVS
  * when using the mouse to set the cursor position, the color of the
    character under the cursor is now saved, set, and restored correctly
  * minor utility and Windows installer improvements and fixes
  * some optimizations in software video mode, and serial floppy drive
    and printer emulation

Changes in version 1.2.7
------------------------

  * added support for using the mouse to change the cursor position in
    the emulated machine, and copying text to and pasting from the
    clipboard; the following functions are available:
      left click: sets the cursor position
      right click: sets the cursor position, and copies the selected
        word to the clipboard
      ctrl + right click: sets the cursor position, and copies text from
        the selected line to the clipboard
      shift + right click: copies text from the whole screen to the
        clipboard without changing the cursor position

src/cpu.hpp

// plus4emu -- portable Commodore Plus/4 emulator
// Copyright (C) 2003-2017 Istvan Varga <istvanv@users.sourceforge.net>
// https://github.com/istvan-v/plus4emu/
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#ifndef PLUS4EMU_CPU_HPP
#define PLUS4EMU_CPU_HPP

#include "plus4emu.hpp"
#include "bplist.hpp"

namespace Plus4 {

  class M7501Registers {
   public:
    uint16_t  reg_PC;
    uint8_t   reg_SR;
    uint8_t   reg_AC;
    uint8_t   reg_XR;
    uint8_t   reg_YR;
    uint8_t   reg_SP;
    M7501Registers()
      : reg_PC(0x0000),
        reg_SR(uint8_t(0x24)),
        reg_AC(0x00),
        reg_XR(0x00),
        reg_YR(0x00),
        reg_SP(uint8_t(0xFF))
    {
    }
  };

  typedef PLUS4EMU_REGPARM2 uint8_t (*M7501MemoryReadCallback)(void *userData,
                                                               uint16_t addr);

  typedef PLUS4EMU_REGPARM3 void (*M7501MemoryWriteCallback)(void *userData,
                                                             uint16_t addr,
                                                             uint8_t value);

  class M7501 : protected M7501Registers {
   private:
    static const unsigned char  CPU_OP_RD_OPCODE            =  0;
    static const unsigned char  CPU_OP_RD_TMP               =  1;
    static const unsigned char  CPU_OP_RD_TMP_NODEBUG       =  2;
    static const unsigned char  CPU_OP_RD_L                 =  3;
    static const unsigned char  CPU_OP_RD_H                 =  4;
    static const unsigned char  CPU_OP_LD_TMP_MEM           =  5;
    static const unsigned char  CPU_OP_LD_MEM_TMP           =  6;
    static const unsigned char  CPU_OP_LD_MEM_TMP_NODEBUG   =  7;
    static const unsigned char  CPU_OP_LD_H_MEMP1_L_TMP     =  8;
    static const unsigned char  CPU_OP_LD_DUMMY_MEM_PC      =  9;
    static const unsigned char  CPU_OP_LD_DUMMY_MEM_SP      = 10;
    static const unsigned char  CPU_OP_PUSH_TMP             = 11;
    static const unsigned char  CPU_OP_POP_TMP              = 12;
    static const unsigned char  CPU_OP_PUSH_PCL             = 13;
    static const unsigned char  CPU_OP_POP_PCL              = 14;
    static const unsigned char  CPU_OP_PUSH_PCH             = 15;
    static const unsigned char  CPU_OP_POP_PCH              = 16;
    static const unsigned char  CPU_OP_DEC_SP               = 17;
    static const unsigned char  CPU_OP_LD_TMP_SR            = 18;
    static const unsigned char  CPU_OP_LD_TMP_A             = 19;
    static const unsigned char  CPU_OP_LD_PC_HL             = 20;
    static const unsigned char  CPU_OP_LD_SR_TMP            = 21;
    static const unsigned char  CPU_OP_LD_A_TMP             = 22;
    static const unsigned char  CPU_OP_ADDR_X               = 23;
    static const unsigned char  CPU_OP_ADDR_X_SLOW          = 24;
    static const unsigned char  CPU_OP_ADDR_X_ZEROPAGE      = 25;
    static const unsigned char  CPU_OP_ADDR_X_SHY           = 26;
    static const unsigned char  CPU_OP_ADDR_Y               = 27;
    static const unsigned char  CPU_OP_ADDR_Y_SLOW          = 28;
    static const unsigned char  CPU_OP_ADDR_Y_ZEROPAGE      = 29;
    static const unsigned char  CPU_OP_ADDR_Y_SHA           = 30;
    static const unsigned char  CPU_OP_ADDR_Y_SHS           = 31;
    static const unsigned char  CPU_OP_ADDR_Y_SHX           = 32;
    static const unsigned char  CPU_OP_SET_NZ               = 33;
    static const unsigned char  CPU_OP_ADC                  = 34;
    static const unsigned char  CPU_OP_ANC                  = 35;
    static const unsigned char  CPU_OP_AND                  = 36;
    static const unsigned char  CPU_OP_ANE                  = 37;
    static const unsigned char  CPU_OP_ARR                  = 38;
    static const unsigned char  CPU_OP_ASL                  = 39;
    static const unsigned char  CPU_OP_BCC                  = 40;
    static const unsigned char  CPU_OP_BCS                  = 41;
    static const unsigned char  CPU_OP_BEQ                  = 42;
    static const unsigned char  CPU_OP_BIT                  = 43;
    static const unsigned char  CPU_OP_BMI                  = 44;
    static const unsigned char  CPU_OP_BNE                  = 45;
    static const unsigned char  CPU_OP_BPL                  = 46;
    static const unsigned char  CPU_OP_BRK                  = 47;
    static const unsigned char  CPU_OP_BVC                  = 48;
    static const unsigned char  CPU_OP_BVS                  = 49;
    static const unsigned char  CPU_OP_CLC                  = 50;
    static const unsigned char  CPU_OP_CLD                  = 51;
    static const unsigned char  CPU_OP_CLI                  = 52;
    static const unsigned char  CPU_OP_CLV                  = 53;
    static const unsigned char  CPU_OP_CMP                  = 54;
    static const unsigned char  CPU_OP_CPX                  = 55;
    static const unsigned char  CPU_OP_CPY                  = 56;
    static const unsigned char  CPU_OP_DEC                  = 57;
    static const unsigned char  CPU_OP_DEX                  = 58;
    static const unsigned char  CPU_OP_DEY                  = 59;
    static const unsigned char  CPU_OP_EOR                  = 60;
    static const unsigned char  CPU_OP_INC                  = 61;
    static const unsigned char  CPU_OP_INTERRUPT            = 62;
    static const unsigned char  CPU_OP_INX                  = 63;
    static const unsigned char  CPU_OP_INY                  = 64;
    static const unsigned char  CPU_OP_JMP_RELATIVE         = 65;
    static const unsigned char  CPU_OP_LAS                  = 66;
    static const unsigned char  CPU_OP_LAX                  = 67;
    static const unsigned char  CPU_OP_LDA                  = 68;
    static const unsigned char  CPU_OP_LDX                  = 69;
    static const unsigned char  CPU_OP_LDY                  = 70;
    static const unsigned char  CPU_OP_LSR                  = 71;
    static const unsigned char  CPU_OP_LXA                  = 72;
    static const unsigned char  CPU_OP_ORA                  = 73;
    static const unsigned char  CPU_OP_RESET                = 74;
    static const unsigned char  CPU_OP_ROL                  = 75;
    static const unsigned char  CPU_OP_ROR                  = 76;
    static const unsigned char  CPU_OP_SAX                  = 77;
    static const unsigned char  CPU_OP_SBC                  = 78;
    static const unsigned char  CPU_OP_SBX                  = 79;
    static const unsigned char  CPU_OP_SEC                  = 80;
    static const unsigned char  CPU_OP_SED                  = 81;
    static const unsigned char  CPU_OP_SEI                  = 82;
    static const unsigned char  CPU_OP_STA                  = 83;
    static const unsigned char  CPU_OP_STX                  = 84;
    static const unsigned char  CPU_OP_STY                  = 85;
    static const unsigned char  CPU_OP_TAX                  = 86;
    static const unsigned char  CPU_OP_TAY                  = 87;
    static const unsigned char  CPU_OP_TSX                  = 88;
    static const unsigned char  CPU_OP_TXA                  = 89;
    static const unsigned char  CPU_OP_TXS                  = 90;
    static const unsigned char  CPU_OP_TYA                  = 91;
    static const unsigned char  CPU_OP_INVALID_OPCODE       = 92;
    static const unsigned char  opcodeTable[4128];
    const unsigned char *currentOpcode;
    uint8_t     interruptDelayRegister;
    bool        interruptFlag;
    bool        resetFlag;
    bool        haltFlag;
    uint8_t     reg_TMP;
    uint8_t     reg_L;
    uint8_t     reg_H;
    bool        breakOnInvalidOpcode;
    M7501MemoryReadCallback   *memoryReadCallbacks;
    M7501MemoryWriteCallback  *memoryWriteCallbacks;
    void        *memoryCallbackUserData;
    uint8_t     *breakPointTable;
    unsigned int  breakPointCnt;
    // 0: normal mode, 1: single step, 2: step over, 3: trace
    uint8_t     singleStepMode;
    bool        haveBreakPoints;
    uint8_t     breakPointPriorityThreshold;
    int32_t     singleStepModeNextAddr;
    int32_t     newPCAddress;
    void checkOpcodeReadBreakPoint(uint16_t addr, uint8_t value);
    void checkReadBreakPoint(uint16_t addr, uint8_t value);
    void checkWriteBreakPoint(uint16_t addr, uint8_t value);
   protected:
    inline uint8_t readMemory(uint16_t addr)
    {
      return (memoryReadCallbacks[addr](memoryCallbackUserData, addr));
    }
    inline void writeMemory(uint16_t addr, uint8_t value)
    {
      memoryWriteCallbacks[addr](memoryCallbackUserData, addr, value);
    }
   public:
    M7501();
    virtual ~M7501();
    inline void setMemoryReadCallback(uint16_t addr_,
                                      M7501MemoryReadCallback func)
    {
      memoryReadCallbacks[addr_] = func;
    }
    inline void setMemoryWriteCallback(uint16_t addr_,
                                       M7501MemoryWriteCallback func)
    {
      memoryWriteCallbacks[addr_] = func;
    }
    inline void setMemoryCallbackUserData(void *userData)
    {
      memoryCallbackUserData = userData;
    }
    inline M7501MemoryReadCallback getMemoryReadCallback(uint16_t addr_) const
    {
      return memoryReadCallbacks[addr_];
    }
    inline M7501MemoryWriteCallback getMemoryWriteCallback(uint16_t addr_) const
    {
      return memoryWriteCallbacks[addr_];
    }
    void runOneCycle_RDYHigh();
    void runOneCycle_RDYLow();
    inline void runOneCycle()
    {
      if (PLUS4EMU_EXPECT(!haltFlag))
        runOneCycle_RDYHigh();
      else
        runOneCycle_RDYLow();
    }
    inline void run_RDYHigh(int nCycles = 1)
    {
      do {
        this->runOneCycle_RDYHigh();
      } while (--nCycles);
    }
    inline void run_RDYLow(int nCycles = 1)
    {
      do {
        this->runOneCycle_RDYLow();
      } while (--nCycles);
    }
    inline void run(int nCycles = 1)
    {
      if (PLUS4EMU_EXPECT(!haltFlag))
        run_RDYHigh(nCycles);
      else
        run_RDYLow(nCycles);
    }
    inline void interruptRequest()
    {
      // delay interrupt requests by 2 cycles
      interruptDelayRegister |= uint8_t(0x04);
    }
    inline void clearInterruptRequest()
    {
      interruptDelayRegister &= uint8_t(0x03);
    }
    inline void interruptRequest(bool isIRQ)
    {
      // delay interrupt requests by 2 cycles
      interruptDelayRegister &= uint8_t(0x03);
      interruptDelayRegister |= (uint8_t(isIRQ) << 2);
    }
    virtual void reset(bool isColdReset = false);
    inline void setIsCPURunning(bool n)
    {
      haltFlag = !n;
    }
    inline bool getIsCPURunning() const
    {
      return (!haltFlag);
    }
    inline void setOverflowFlag()
    {
      reg_SR |= uint8_t(0x40);
    }
    // 'type' can be the sum of any of:
    //   1: memory read
    //   2: memory write
    //   4: execute (opcode read)
    //   8: ignore
    // a negative priority value deletes a previously set breakpoint
    void setBreakPoint(int type, uint16_t addr, int priority);
    void clearBreakPoints();
    void setBreakPointPriorityThreshold(int n);
    inline int getBreakPointPriorityThreshold() const
    {
      return int(breakPointPriorityThreshold >> 4);
    }
    Plus4Emu::BreakPointList getBreakPointList();
    // 'mode_' can be one of the following values:
    //   0: normal mode
    //   1: single step (break on every instruction, disable breakpoints)
    //   2: step over (skip JSR or branch instruction)
    //   3: trace (break on every instruction, breakpoints are not disabled)
    //   4: step into (break after branch instruction only if branch is taken)
    void setSingleStepMode(int mode_);
    // Set the next address where single step mode will stop, ignoring any
    // other instructions. If 'addr' is negative, then a break is triggered
    // immediately at the next instruction.
    // Note: setSingleStepMode() must be called first with a mode parameter
    // of 2 or 4.
    void setSingleStepModeNextAddress(int32_t addr);
    inline void setBreakOnInvalidOpcode(bool isEnabled)
    {
      this->breakOnInvalidOpcode = isEnabled;
    }
    inline bool getIsBreakOnInvalidOpcode() const
    {
      return (this->breakOnInvalidOpcode);
    }
    void setRegisters(const M7501Registers& r);
    void getRegisters(M7501Registers& r) const;
    void saveState(Plus4Emu::File::Buffer&);
    void saveState(Plus4Emu::File&);
    void loadState(Plus4Emu::File::Buffer&);
    void registerChunkType(Plus4Emu::File&);
   protected:
    // type is 0 for opcode read at breakpoint, 1 for memory read,
    // 2 for memory write, and 3 for opcode read in single step mode
    virtual void breakPointCallback(int type, uint16_t addr, uint8_t value);
  };

}       // namespace Plus4

#endif  // PLUS4EMU_CPU_HPP

